gapic/samplegen/samplegen.py

```diff
@@ -102,6 +104,81 @@ class TransformedRequest:
     base: str
     single: Optional[AttributeRequestSetup]
     body: Optional[List[AttributeRequestSetup]]
+    pattern: Optional[str] = None
+
+    # Resource patterns look something like
+    # kingdom/{kingdom}/phylum/{phylum}/class/{class}
+    RESOURCE_RE = re.compile(r"\{([^}/]+)\}")
+
+    @classmethod
+    def build(cls, request_type, api_schema, base, attrs, is_resource_request):
+        """Build a TransformedRequest based on parsed inpu
+
+        Acts as a factory to hide complicated logic for resource based requests.
+
+        Args:
+            request_type (wrappers.MessageType): The method's request message type.
+            api_schema (api.API): The API schema (used for looking up other messages)
+            base (str): the name of the base field being set.
+            attrs [AttributeRequestSetup]: All the attributes (or fields) being set
+                                           for the base field.
+            is_resource_request (bool): Indicates whether the request describes a
+                                        constructed resource name path.
+        """
+
+        if not attrs[0].field:
+            return cls(base=base, body=None, single=attrs[0])
+        elif not is_resource_request:
+            return cls(base=base, body=attrs, single=None)
+        else:
+            # This is the tricky one.
+            # We need to determine whether the field is describing a valid resource,
+            # and if so, what its corresponding message type is.
+            # Then we need to find the pattern with parameters
+            # that exatcly matches the attrs, if one exists.
+            #
+            # It's a precondition that the base field is
+            # a valid field of the request message type.
+            resource_typestr = (request_type.
+                                fields[base].
+                                options.
+                                Extensions[resource_pb2.resource_reference].
+                                type)
+
+            resource_message_descriptor = next(
+                (msg.options.Extensions[resource_pb2.resource]
+                 for msg in api_schema.messages.values()
+                 if msg.options.Extensions[resource_pb2.resource].type == resource_typestr),
+                None
+            )
+            if not resource_message_descriptor:
+                raise types.NoSuchResource(
+                    f"No message exists for resource: {resource_typestr}")
+
+            attr_names = [attr.field for attr in attrs]
+
+            # A single resource may be found under multiple paths and have many patterns.
+            # We want to find an _exact_ match, if one exists.
+            pattern = next((p
+                            for p in resource_message_descriptor.pattern
+                            if cls.RESOURCE_RE.findall(p) == attr_names), None)
```

Comment:
Would it be possible that `attr_names` has different order than `resource_message_descriptor.pattern`, and the search would find nothing?

Reply:
There's nothing in the samplegen spec about ordering requirements for resource-name based request construction. Allowing it would be tricky but doable; my concern is that there may be multiple valid paths that use the same set of parameters but in different order. I don't know if that's common or even allowed, but it was my concern.

Reply:
Oh are you concerned about there being two patterns like the following at the same time:
- `foos/{foo}/bars/{bar}`
- `bars/{bar}/foos/{foo}`

I would say the possibility that there is such a resource name is very unlikely, and allowing unordered input should prevail that. However I don't think a sample config with reversed order of resource name entities would be very likely either. So feel free to leave the implementation as it is until (if possible at all) someone complains :)

Reply:
Exactly; my feeling as well is to be strict until someone complains :P